### ISSUE TITLE
Issue with bitwise operators 

### ISSUE THREAD

**Issue by acope3** [riboviz/riboviz#214]
Hi everyone,

I identified a bug in riboviz that occurs when processing the multiplexed files. Although I'm successfully demultiplexing the multiplexed fastq file into separate fastq files, only a few of these demultiplexed fastq files are actually getting processed. This has to do with `get_non_zero_deplexed_samples` in `sample_sheets.py`. 

The code currently looks like
`non_zero_samples = sample_sheet[~sample_sheet[SAMPLE_ID].isin([UNASSIGNED_TAG, TOTAL_READS]) & sample_sheet[NUM_READS] != 0]`

However, because & is a bitwise operator, this isn't behaving as we expect. We want to put each condition into its own set of parentheses. As an example, see the attached screenshot. Although I'm pretty confident I've identified the issue, I'm bringing it to everyone's attention in case there are other places in the code which might have this issue. 
![image](https://user-images.githubusercontent.com/21202517/93799778-9f8fd780-fc0d-11ea-91ee-90a0c0ce805f.png)

**Comment by mikej888** [comment]
`&` wasn't intended as a bitwise AND. I think this may arise from Pandas data frames still being new to me when I wrote that. It should be a logical AND.

**Comment by acope3** [comment]
> `&` wasn't intended as a bitwise AND. I think this may arise from Pandas data frames still being new to me when I wrote that. It should be a logical AND.

I don't think pandas data frames work with logical AND. I've always had to use the bitwise AND.

**Comment by mikej888** [comment]
Argh, so much for consistency. I'm writing a test class for this case just now.

**Comment by mikej888** [comment]
Please could you post the sample sheet that gave that behaviour? Thanks

**Comment by acope3** [comment]
Here is the sample sheet.
[GSE124428_barcode.txt](https://github.com/riboviz/riboviz/files/5268805/GSE124428_barcode.txt)

**Comment by mikej888** [comment]
Sorry, do you have the one with the NumReads column too and the Total and Unassigned columns.

**Comment by acope3** [comment]
[num_reads.txt](https://github.com/riboviz/riboviz/files/5268840/num_reads.txt)

**Comment by mikej888** [comment]
I've commited a test, `riboviz.test.test_sample_sheet` to the `bitwise-sample-sheet-214` branch. This test can be run as follows:

```console
$ pytest -vs riboviz/test/test_sample_sheet.py
```

This includes a test, `test_get_non_zero_deplexed_samples`, that is designed to fail on the current, flawed, implementation of `get_non_zero_deplexed_samples`. The test is based on your sample sheet data and once you commit the fix to `get_non_zero_deplexed_samples` the test should pass.

**Comment by mikej888** [comment]
The only other occurrences of bitwise `&` being used to filter Pandas data frames are in branches `gff-spec-74` and `orf-finder-92-mikej888` (which is kept in sync with `gff-spec-74`). This occurrence is in `riboviz/test/test_fasta_gff.py` which already has parentheses to ensure `&` is applied correctly:

```
        issue_df = df[(df[fasta_gff.SEQUENCE] == sequence) &
                      (df[fasta_gff.FEATURE] == feature) &
                      (df[fasta_gff.ISSUE] == issue)]
```

**Comment by acope3** [comment]
@mikej888 Did you push your fixes to the repository already? Can we close this issue?

**Comment by mikej888** [comment]
@acope3 I thought, when we discussed it, that I was to write a test that would fail, you would apply the fix so the test passes then we can merge into `develop` and we'd be done (see [comment](https://github.com/riboviz/riboviz/issues/214#issuecomment-698226530) above)). I'd implemented the test.

If you just want me to go ahead and do the fix as well them please let me know.

**Comment by acope3** [comment]
Submitted pull request #223.

**Comment by mikej888** [comment]
Merged into 'develop' via pull request #223 into branch `bitwise-sample-sheet-214` and pull request #224 into `develop`.
